lib/utils.js

```diff
@@ -1,73 +1,134 @@
+}
+
+async function getXcodeTraceTemplatePath () {
+  let {stdout} = await exec('xcode-select', ['-print-path']);
+  let xCodePath = stdout.trim().replace('\n$', '');
+  for(let ext of ['bundle', 'xrplugin']) {
+    let traceTemplatePath = path.resolve(xCodePath, '..', '..',
+      'Contents/Applications/Instruments.app',
+      `Contents/PlugIns/AutomationInstrument.${ext}/Contents/Resources/` +
+        'Automation.tracetemplate');
+    let fd;
+    try {
+      fd = await fs.open(traceTemplatePath, 'r');
       return traceTemplatePath;
-    });
-};
+    } catch (ign) {}
+    finally {
+      try { if (fd) { await fs.close(fd); } } catch (ign) {}
+    }
+  }
+  throw new Error('Could not retrieve trace template!');
+}
 
-exports.getXcodeTraceTemplatePath = getXcodeTraceTemplatePath;
+async function getInstrumentsPath () {
+  let instrumentsPath;
+  try {
+    let {stdout} = await exec('xcrun', ['-find', 'instruments']);
+    if (typeof stdout === "undefined") stdout = "";
+    instrumentsPath = stdout.trim().replace('\n$', '');
+  } catch (err) {
+    if (err) log.error(err.message);
+  }
+  if (!instrumentsPath) {
+    log.error("Could not find instruments binary");
+    throw new Error("Could not find the instruments " +
+        "binary. Please ensure `xcrun -find instruments` can locate it.");
+  }
+  log.debug("Instruments is at: " + instrumentsPath);
+  return instrumentsPath;
+}
 
-exports.killAllSimulators = function () {
-  return exec('`which pkill` -f iPhoneSimulator').catch(function () {})
-    .delay(1000);
-};
+async function getAvailableDevices () {
+  log.debug("Getting list of devices instruments supports");
+  let instrumentsPath = await getInstrumentsPath();
+  let opts = {timeout: INST_STALL_TIMEOUT};
+  let lines;
+  try {
+    let {stdout} = await exec(instrumentsPath, ['-s', 'devices'], opts);
+    lines = stdout.split("\n");
+  } catch (err) {
+    log.error(err);
+    throw new Error(`Failed getting devices, err: ${err}.`);
+  }
+  let devices = [];
+  for(let line of lines) {
+    if (/^i.+$/.test(line)) {
+      devices.push(line);
+    }
+  }
+  return devices;
+}
 
-exports.killAllInstruments = function () {
-  return exec('`which pkill` -f instruments').catch(function () {})
-    .delay(1000);
-};
 
-exports.cleanAllTraces = function () {
+// TODO: use simulator package
+async function killAllSimulators () {
+  let version = await xcode.getVersion();
+  // this is contrary to our usual pattern, but if getting the xcode version
+  // fails, we couldn't have started simulators anyways/
+  log.debug("Killing all iOS Simulator");
+  try {
+    await exec('pkill', ['-9', '-f',
+      version >= "6" ? 'iOS Simulator' : 'iPhoneSimulator']);
+  } catch (ign) {}
+}
+
+async function killAllInstruments () {
+  log.debug("Killing all instruments");
+  try {
+    await exec('pkill',  ['-f', 'instruments']);
+  } catch (ign) {}
+}
+
+async function cleanAllTraces () {
   if (process.env.CLEAN_TRACES) {
```

Comment:
shouldn't we put this check around where we _call_ `cleanAllTraces` instead of inside this method?